devise課題の提出物

yocaji opened the pull request:
devise課題の提出物になります。レビューよろしくお願いいたします🙏

### プラクティスの参考ページおよび公式情報以外で参考にした記事等

- ログアウト時のリダイレクト先を`/books`に変更する
  - [【devise】アカウント登録、ログイン/ ログアウト、アカウント編集後のリダイレクト先の変更 - qiita](https://qiita.com/tatty/items/a9759755e562ac4693ec)
- 登録情報に郵便番号・住所・自己紹介文を追加する
  - [[rails]deviseにカラムを追加する - qiita](https://qiita.com/nowshi/items/a7220036a684a9ac4bb4)
  - [deviseでストロングパラメーターを使えるようにする - qiita](https://qiita.com/mr0216/items/488b85d031b10913d74b)
- ユーザーの一覧画面と詳細画面を作成する
  - [devise で作成した user モデル用のコントローラーの index, show アクションを追加 | easyramble](https://easyramble.com/create-users-index-show-on-devise.html)
  - [list all users and user profile with devise in rails | by wajeeh ahsan | medium](https://wajeeh-ahsan.medium.com/list-all-users-and-user-profile-with-devise-in-rails-d5e563bb1e48)

Commit:
deviseをセットアップした

Commit:
devise-i18nを入れた

Commit:
letter_opener_webをセットアップした

Commit:
登録情報の項目を追加した

Commit:
ログアウト後のリダイレクト先を変更した

Commit:
ユーザー一覧ページとユーザー詳細ページを作成した

Commit:
日本語化対応

Commit:
rubocopの指摘に対応

Commit:
ダミーユーザー生成のseedsを追加

Commit:
誤って削除したと思われる行を復元した

Commit:
不要なfixturesを削除した

JunichiIto reviewed (commented):
動画でレビューしたので別途リンクを送ります〜。

JunichiIto commented on app/views/devise/registrations/edit.html.erb:
```diff
@@ -0,0 +1,58 @@
+
+  <div class="field">
+    <%= f.label :postcode %><br />
+    <%= f.text_field :postcode, autofocus: true, autocomplete: "postcode" %>
```
動画では話し忘れましたが、「`autofocus`もなんで付けるの？付ける意味あるの？」というのを自問自答してみてほしいです〜。

yocaji replied:
`autofocus:`はhtmlタグの`autofocus`を実現するものなので、1画面に複数指定するのは意味がないですね😨
既存のemailのフォームにフォーカスが当たれば良いので、追加した項目の`aurofocus:`は削除しました。

`autocomplete:`についても、ブラウザの入力補助を活用するためのものということ、認識不足でした。
- 郵便番号は`postal-code`に修正
- 住所はまとめて取れる項目が無かったため`autocomplete:`は削除
- 自己紹介に設定するのは適切ではないため`autocomplete:`は削除

Commit:
レビューの指摘事項に対応した

- 自動生成された空のファイルを削除
- seed.rbのpassword_confirmation:を削除
- roots.rbのハッシュの記法を修正
- /usersの戻るボタンを削除
- ログアウト時の遷移先としてログイン画面を指定
- rootとして/booksを指定
- 自己紹介の表示にsimple_formatを使わないよう修正
- autofocus, autocompleteの指定を見直した
- devise-i18n/rails/locales/ja.ymlと重複する翻訳は削除して置き換えた

yocaji commented on app/controllers/application_controller.rb:
```diff
@@ -1,4 +1,16 @@
+  before_action :configure_permitted_parameters, if: :devise_controller?
+
+  def after_sign_out_path_for(_resource)
+    :new_user_session
```
ログアウト後の遷移先をログイン画面に変更しました

yocajii commented on app/views/devise/registrations/edit.html.erb:
```diff
@@ -0,0 +1,58 @@
+
+  <div class="field">
+    <%= f.label :postcode %><br />
+    <%= f.text_field :postcode, autocomplete: "postal-code" %>
```
`autofocus:`はhtmlタグの`autofocus`を実現するものなので、1画面に複数指定するのは意味がないですね😨
既存のemailのフォームにフォーカスが当たれば良いので、追加した項目の`aurofocus:`は削除しました。

`autocomplete:`についても、ブラウザの入力補助を活用するためのものということ、認識不足でした。
- 郵便番号は`postal-code`に修正
- 住所はまとめて取れる項目が無かったため`autocomplete:`は削除
- 自己紹介に設定するのは適切ではないため`autocomplete:`は削除

yocaji commented on app/views/layouts/application.html.erb:
```diff
@@ -10,6 +10,24 @@
+        <%= link_to t('labels.user.logout'), destroy_user_session_path, method: :delete, class: 'navbar-link'  %>
+      <% else %>
+        <%= link_to t('devise.shared.links.sign_up'), new_user_registration_path, class: 'navbar-link'  %> |
+        <%= link_to t('devise.shared.links.sign_in'), new_user_session_path, class: 'navbar-link'  %>
```
リンクのテキストを`devise-i18n`の項目を利用するように置き換えました

yocaji commented on app/views/users/index.html.erb:
```diff
@@ -0,0 +1,25 @@
+  </tbody>
+</table>
+
+<%= paginate @users %>
```
`users`と`books`は独立した機能と捉えることにして、戻るボタンは削除しました。

yocaji commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,23 @@
+
+<p>
+  <strong><%= User.human_attribute_name(:self_introduction) %>:</strong>
+  <%= sanitize @user.self_introduction.gsub(/(\r\n|\r|\n)/, "<br>"), tags: ['br'] %>
```
改行コードを置換した上でsanitaizeで<br>のみ許可するようにしました
simlpe_formatの状態で生成されたソースを確認すると、ご指摘の通り`<p>`タグが入れ子になっていました😱

JunichiIto replied:
正規表現の部分、`\R`っていうメタ文字が使えるみたいですよ。

https://qiita.com/akicho8/items/efe59578f12d6b7f5626

yocaji commented on config/locales/translation_ja.yml:
```diff
@@ -9,3 +9,12 @@ ja:
+    user:
+      show: ユーザー詳細
+      edit: ユーザー編集
+      logout: ログアウト
```
ユーザー登録とログインの項目は`devise-i18n`の方にあったため削除しました

yocajii commented on config/routes.rb:
```diff
@@ -1,4 +1,7 @@
 Rails.application.routes.draw do
+  devise_for :users
+  root 'books#index'
```
ログアウト後に遷移したログイン画面からログインするとrailsの初期画面に遷移してしまうのを避けるため、rootにbooksを指定しました

yocaji commented on config/routes.rb:
```diff
@@ -1,4 +1,7 @@
+  devise_for :users
+  root 'books#index'
   resources :books
+  resources :users, only: [:index, :show]
```
ここだけ古いハッシュ記法になっていたのを修正しました

yocaji commented on db/migrate/20220403103419_add_self_introduction_to_user.rb:
```diff
@@ -0,0 +1,7 @@
+
+class AddSelfIntroductionToUser < ActiveRecord::Migration[6.1]
+  def change
+    add_column :users, :self_introduction, :text
```
動画で見せていただいた公式ドキュメントなどを確認して、SQLiteではstringとしてもtextとしても最終的には同じ動作になっていることを理解しました。
DBの文脈以外でも、stringとtextの違いには注意を払う必要がある事が分かりました🙆‍♂️

今回は自己紹介が長文を受け入れられる項目であることを分かりやすくするため、データ型をtextに変更しました。

yocaji commented on db/seeds.rb:
```diff
@@ -42,4 +42,22 @@ def picture_file(name)
 
+User.create!(
+  email: 'test@example.com',
+  password: 'test123',
```
`password_confirmation:`はご指摘の通り不要だったため削除しました

Commit:
ログイン後の遷移先をafter_sign_in_path_forで指定するようにした

JunichiIto reviewed (approved):
1点コメントしましたが、大きな問題では無いのでこれでOKです〜 🙆‍♂️

Commit:
改行コードにマッチする正規表現を修正
